File: scripts/algorithms/C/Closest Room/Closest Room.java
// Runtime: 164 ms (Top 39.53%) | Memory: 127.2 MB (Top 83.72%)
class Solution {
    public int[] closestRoom(int[][] rooms, int[][] queries) {
        int n = rooms.length, k = queries.length;
        Integer[] indexes = new Integer[k];
        for (int i = 0; i < k; i++) indexes[i] = i;
        Arrays.sort(rooms, (a, b) -> Integer.compare(b[1], a[1])); //Sort by decreasing order of room size
        Arrays.sort(indexes, (a, b) -> Integer.compare(queries[b][1], queries[a][1])); // Sort by decreasing order of query minSize
        TreeSet<Integer> roomIdsSoFar = new TreeSet<>();
        int[] ans = new int[k];
        int i = 0;
        for (int index : indexes) {
            while (i < n && rooms[i][1] >= queries[index][1]) { // Add id of the room which its size >= query minSize
                roomIdsSoFar.add(rooms[i++][0]);
            }
            ans[index] = searchClosetRoomId(roomIdsSoFar, queries[index][0]);
        }
        return ans;
    }
    int searchClosetRoomId(TreeSet<Integer> treeSet, int preferredId) {
        Integer floor = treeSet.floor(preferredId);
        Integer ceiling = treeSet.ceiling(preferredId);
        int ansAbs = Integer.MAX_VALUE, ans = -1;
        if (floor != null) {
            ans = floor;
            ansAbs = Math.abs(preferredId - floor);
        }
        if (ceiling != null && ansAbs > Math.abs(preferredId - ceiling)) {
            ans = ceiling;
        }
        return ans;
    }
}

File: scripts/algorithms/C/Closest Room/Closest Room.cpp
// Runtime: 1335 ms (Top 26.39%) | Memory: 149.2 MB (Top 71.30%)
class Solution {
public:
    vector<int> closestRoom(vector<vector<int>>& rooms, vector<vector<int>>& queries) {
        const int n = rooms.size(), m = queries.size();
        for(int i = 0; i < m; i++) queries[i].push_back(i);
        sort(begin(queries), end(queries), [](const auto& a, const auto& b){return a[1] > b[1];});
        sort(begin(rooms), end(rooms), [](const auto& a, const auto& b){return a[1] > b[1];});
        vector<int> ans(m, -1);
        set<int> ids;
        int j = 0;
        for(const auto& q: queries){
            while(j < n && rooms[j][1] >=q[1]) ids.insert(rooms[j++][0]);
            if(ids.empty()) continue;
            int id = q[0];
            auto it = ids.lower_bound(id);
            int id1 = (it != end(ids) )? *it : INT_MAX;
            int id2 = id1;
            if(it != begin(ids)) id2 = *prev(it);
            ans[q[2]] = abs(id1 - id) < abs(id2 - id) ? id1 : id2;
        }
        return ans;
    }
};